README.md
[![Build status](https://travis-ci.org/ccmikechen/videx.svg?branch=master)](https://travis-ci.org/ccmikechen/videx)
[![Videx version](https://img.shields.io/hexpm/v/videx.svg)](https://hex.pm/packages/videx)
[![Hex.pm](https://img.shields.io/hexpm/dt/videx.svg)](https://hex.pm/packages/videx)
[![Inline docs](https://inch-ci.org/github/ccmikechen/videx.svg?branch=master)](https://inch-ci.org/github/ccmikechen/videx)

# Videx

Videx is a simple video url parser for parsing and generating video urls.

## Usage

Parse Youtube video url.

```elixir
Videx.parse("https://www.youtube.com/watch?v=Hh9yZWeTmVM")
# %Videx.Youtube{id: "Hh9yZWeTmVM", params: %{}, type: :long}
```

Generate Youtube video url.

```elixir
video = %Videx.Youtube{id: "Hh9yZWeTmVM"}
Videx.Youtube.generate(video, :long)
# "https://www.youtube.com/watch?v=Hh9yZWeTmVM"
```

Here is how you can convert Youtube video url to another format.

```elixir
"https://www.youtube.com/watch?v=Hh9yZWeTmVM"
|> Videx.parse()
|> Videx.Youtube.generate(:short)
# "https://youtu.be/Hh9yZWeTmVM"
```

## Installation

Add Videx to your `mix.exs`:

```elixir
def deps do
  [
    {:videx, "~> 0.2.0"}
  ]
end
```
After that, run `mix deps.get`.

## Features

Current Features or To-Do

- [x] Supports: parse and generate url for
  - [x] [YouTube](https://www.youtube.com/)
  - [x] [Vimeo](https://vimeo.com/)
  - [ ] [Twitch](https://www.twitch.tv/)
  - [ ] [SoundCloud](https://soundcloud.com/)
  - [ ] [Youku](https://www.youku.com/)
  - [ ] [Coub](https://coub.com/)
  - [ ] [Wistia](https://wistia.com/)

## License

Videx is under MIT license. Check the LICENSE file for more details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2019-08-16
### Added
- `html/1` and `html/2` to Youtube to generate embed HTML.
- Vimeo support.

## [0.1.0] - 2019-08-15
### Added
- `parse/1` and `generate/3` for Youtube.
- CHANGELOG.md following the [Keep a changelog](http://keepachangelog.com/).

[Unreleased]: https://github.com/ccmikechen/videx/compare/v0.2.0...HEAD
[0.2.0]: https://github.com/ccmikechen/videx/releases/tag/v0.2.0
[0.1.0]: https://github.com/ccmikechen/videx/releases/tag/v0.1.0


test/videx_test.exs
defmodule VidexTest do
  use ExUnit.Case
  doctest Videx

  @youtube_long "https://www.youtube.com/watch?v=Hh9yZWeTmVM"
  @youtube_short "https://youtu.be/Hh9yZWeTmVM?t=80"
  @youtube_embed "https://www.youtube.com/embed/Hh9yZWeTmVM"
  @youtube_oembed "https://www.youtube.com/oembed?url=https://www.youtube.com/watch?v=Hh9yZWeTmVM&format=json"
  @youtube_video "http://www.youtube.com/v/Hh9yZWeTmVM?version=3&autohide=1"

  @vimeo_short "https://vimeo.com/22439234"
  @vimeo_embed "http://player.vimeo.com/video/22439234?title=0&byline=0"

  @elixir_url "https://elixir-lang.org/"

  test "parse Youtube long URL" do
    parsed = Videx.parse(@youtube_long)
    assert parsed == %Videx.Youtube{type: :long, id: "Hh9yZWeTmVM", params: %{}}
  end

  test "parse Youtube short URL" do
    parsed = Videx.parse(@youtube_short)
    assert parsed == %Videx.Youtube{type: :short, id: "Hh9yZWeTmVM", params: %{"t" => "80"}}
  end

  test "parse Youtube embed URL" do
    parsed = Videx.parse(@youtube_embed)
    assert parsed == %Videx.Youtube{type: :embed, id: "Hh9yZWeTmVM", params: %{}}
  end

  test "parse Youtube oembed URL" do
    parsed = Videx.parse(@youtube_oembed)

    assert parsed == %Videx.Youtube{
             type: :oembed,
             id: "Hh9yZWeTmVM",
             params: %{"format" => "json"}
           }
  end

  test "parse Youtube video URL" do
    parsed = Videx.parse(@youtube_video)

    assert parsed == %Videx.Youtube{
             type: :video,
             id: "Hh9yZWeTmVM",
             params: %{"version" => "3", "autohide" => "1"}
           }
  end

  test "parse Vimeo short URL" do
    parsed = Videx.parse(@vimeo_short)
    assert parsed == %Videx.Vimeo{type: :short, id: "22439234", params: %{}}
  end

  test "parse Vimeo embed URL" do
    parsed = Videx.parse(@vimeo_embed)

    assert parsed == %Videx.Vimeo{
             type: :embed,
             id: "22439234",
             params: %{"title" => "0", "byline" => "0"}
           }
  end

  test "parse non-video URL" do
    parsed = Videx.parse(@elixir_url)
    assert parsed == nil
  end
end


test/videx/vimeo_test.exs
defmodule Videx.VimeoTest do
  use ExUnit.Case
  doctest Videx.Vimeo

  alias Videx.Vimeo

  @video %Vimeo{id: "22439234"}

  test "generate Vimeo short URL" do
    url = Vimeo.generate(@video, :short)
    assert url == "https://vimeo.com/22439234"
  end

  test "generate Vimeo embed URL" do
    url = Vimeo.generate(@video, :embed)
    assert url == "https://player.vimeo.com/video/22439234"
  end

  test "generate Vimeo embed URL with params" do
    url = Vimeo.generate(@video, :embed, %{title: 0, byline: 0})
    assert url == "https://player.vimeo.com/video/22439234?byline=0&title=0"
  end

  test "generate Vimeo embed HTML" do
    html = Vimeo.html(@video)

    assert html ==
             "<iframe src=\"https://player.vimeo.com/video/22439234\" width=\"640\" height=\"360\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen></iframe>"
  end

  test "generate Vimeo embed HTML with params" do
    html =
      Vimeo.html(@video,
        width: 500,
        height: 500,
        color: "ff0000",
        title: false,
        byline: false,
        autoplay: true,
        loop: true
      )

    assert html ==
             "<iframe src=\"https://player.vimeo.com/video/22439234?autoplay=1&byline=0&color=ff0000&loop=1&title=0\" width=\"500\" height=\"500\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen></iframe>"
  end
end


test/videx/youtube_test.exs
defmodule Videx.YoutubeTest do
  use ExUnit.Case
  doctest Videx.Youtube

  alias Videx.Youtube

  @video %Youtube{id: "Hh9yZWeTmVM"}

  test "generate Youtube long URL" do
    url = Youtube.generate(@video, :long)
    assert url == "https://www.youtube.com/watch?v=Hh9yZWeTmVM"
  end

  test "generate Youtube long URL with params" do
    url = Youtube.generate(@video, :long, %{t: 80})
    assert url == "https://www.youtube.com/watch?t=80&v=Hh9yZWeTmVM"
  end

  test "generate Youtube short URL" do
    url = Youtube.generate(@video, :short)
    assert url == "https://youtu.be/Hh9yZWeTmVM"
  end

  test "generate Youtube short URL with params" do
    url = Youtube.generate(@video, :short, %{t: 80})
    assert url == "https://youtu.be/Hh9yZWeTmVM?t=80"
  end

  test "generate Youtube embed URL" do
    url = Youtube.generate(@video, :embed)
    assert url == "https://www.youtube.com/embed/Hh9yZWeTmVM"
  end

  test "generate Youtube embed URL with params" do
    url = Youtube.generate(@video, :embed, %{t: 80})
    assert url == "https://www.youtube.com/embed/Hh9yZWeTmVM?t=80"
  end

  test "generate Youtube oembed URL" do
    url = Youtube.generate(@video, :oembed)

    assert url ==
             "https://www.youtube.com/oembed?url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DHh9yZWeTmVM"
  end

  test "generate Youtube oembed URL with params" do
    url = Youtube.generate(@video, :oembed, %{format: :json})

    assert url ==
             "https://www.youtube.com/oembed?format=json&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DHh9yZWeTmVM"
  end

  test "generate Youtube video URL" do
    url = Youtube.generate(@video, :video)
    assert url == "https://www.youtube.com/v/Hh9yZWeTmVM"
  end

  test "generate Youtube video URL with params" do
    url = Youtube.generate(@video, :video, %{version: 3, autohide: 1})
    assert url == "https://www.youtube.com/v/Hh9yZWeTmVM?autohide=1&version=3"
  end

  test "generate Youtube embed HTML" do
    html = Youtube.html(@video)

    assert html ==
             "<iframe width=\"560\" height=\"315\" src=\"https://www.youtube.com/embed/Hh9yZWeTmVM\" frameborder=\"0\" allow=\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
  end

  test "generate Youtube embed HTML with params" do
    html =
      Youtube.html(@video, width: 100, height: 100, controls: false, nocookie: true, start_at: 100)

    assert html ==
             "<iframe width=\"100\" height=\"100\" src=\"https://www.youtube-nocookie.com/embed/Hh9yZWeTmVM?controls=0&start=100\" frameborder=\"0\" allow=\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
  end
end


mix.exs
defmodule Videx.MixProject do
  use Mix.Project

  @description "Videx is a simple video url parser."
  @version "0.2.0"

  def project do
    [
      app: :videx,
      name: "Videx",
      description: @description,
      version: @version,
      elixir: ">= 1.4.0",
      package: package(),
      deps: deps(),
      source_url: "https://github.com/ccmikechen/videx"
    ]
  end

  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp deps do
    [
      {:ex_doc, "~> 0.21", only: :dev, runtime: false},
      {:inch_ex, ">= 0.0.0", only: :docs}
    ]
  end

  defp package do
    %{
      maintainers: ["Mike Chen"],
      licenses: ["MIT"],
      files: [
        "lib",
        "mix.exs",
        "README.md",
        "LICENSE"
      ],
      links: %{
        "GitHub" => "https://github.com/ccmikechen/videx"
      }
    }
  end
end


lib/videx/vimeo.ex
defmodule Videx.Vimeo do
  use Videx.Video

  @regex ~r/^((?:https?:)?\/\/)?(?:www\.|player\.)?vimeo.com\/(video\/|)(\d+)(?:[a-zA-Z0-9_\-]+)([\w\-]+)(\S+)?/

  @doc """
  Check if the URL is valid video URL for Vimeo.

  ## Examples

      iex> Videx.Vimeo.match?("https://vimeo.com/22439234")
      true

  """
  def match?(url), do: String.match?(url, @regex)

  @doc """
  Parse and extract information from Vimeo video URL.

  ## Examples

      iex> Videx.Vimeo.parse("https://vimeo.com/22439234")
      %Videx.Vimeo{id: "22439234", params: %{}, type: :short}

  """
  def parse(url) do
    case URI.parse(url) do
      %URI{host: "vimeo.com", path: "/" <> id, query: query} ->
        %__MODULE__{type: :short, id: id, params: parse_params(query)}

      %URI{host: "player.vimeo.com", path: "/video/" <> id, query: query} ->
        %__MODULE__{type: :embed, id: id, params: parse_params(query)}

      _ ->
        nil
    end
  end

  @doc """
  Generate URL for Vimeo video.

  ## Examples

      iex> Videx.Vimeo.generate(%Videx.Vimeo{id: "22439234"})
      "https://vimeo.com/22439234"

      iex> Videx.Vimeo.generate("22439234", :embed, %{title: 0})
      "https://player.vimeo.com/video/22439234?title=0"

  """
  def generate(_, type \\ :short, params \\ %{})

  def generate(%__MODULE__{id: id}, :short, params) do
    %URI{
      scheme: "https",
      host: "vimeo.com",
      path: "/#{id}",
      query: encode_query(params)
    }
    |> URI.to_string()
  end

  def generate(%__MODULE__{id: id}, :embed, params) do
    %URI{
      scheme: "https",
      host: "player.vimeo.com",
      path: "/video/#{id}",
      query: encode_query(params)
    }
    |> URI.to_string()
  end

  def generate(id, type, params) when is_binary(id) do
    generate(%__MODULE__{id: id}, type, params)
  end

  @doc """
  Generate embed HTML for Vimeo video.

  ## Examples

      iex> Videx.Vimeo.html(%Videx.Vimeo{id: "22439234"})
      "<iframe src=\\"https://player.vimeo.com/video/22439234\\" width=\\"640\\" height=\\"360\\" frameborder=\\"0\\" allow=\\"autoplay; fullscreen\\" allowfullscreen></iframe>"

      iex> Videx.Vimeo.html("22439234", width: 320, height: 240)
      "<iframe src=\\"https://player.vimeo.com/video/22439234\\" width=\\"320\\" height=\\"240\\" frameborder=\\"0\\" allow=\\"autoplay; fullscreen\\" allowfullscreen></iframe>"

  """
  def html(_, params \\ [])

  def html(%__MODULE__{id: id}, params) do
    width = params[:width] || 640
    height = params[:height] || 360
    color = params[:color]
    portrait = is_nil(params[:portrait]) || params[:portrait]
    title = is_nil(params[:title]) || params[:title]
    byline = is_nil(params[:byline]) || params[:byline]
    autoplay = params[:autoplay]
    loop = params[:loop]

    query =
      %{}
      |> Map.merge(
        if color do
          %{color: color}
        else
          %{}
        end
      )
      |> merge_query(:portrait, portrait, false)
      |> merge_query(:title, title, false)
      |> merge_query(:byline, byline, false)
      |> merge_query(:autoplay, autoplay, true)
      |> merge_query(:loop, loop, true)
      |> encode_query()

    url =
      %URI{
        scheme: "https",
        host: "player.vimeo.com",
        path: "/video/#{id}",
        query: query
      }
      |> URI.to_string()

    "<iframe src=\"#{url}\" width=\"#{width}\" height=\"#{height}\" frameborder=\"0\" allow=\"autoplay; fullscreen\" allowfullscreen></iframe>"
  end

  def html(id, params) when is_binary(id) do
    html(%__MODULE__{id: id}, params)
  end
end


lib/videx/youtube.ex
defmodule Videx.Youtube do
  use Videx.Video

  # From https://stackoverflow.com/a/37704433
  @regex ~r/^((?:https?:)?\/\/)?((?:www|m)\.)?((?:youtube\.com|youtu.be))(\/(?:[\w\-]+\?v=|embed\/|v\/)?)([\w\-]+)(\S+)?$/

  @doc """
  Check if the URL is valid video URL for Youtube.

  ## Examples

      iex> Videx.Youtube.match?("https://www.youtube.com/watch?v=Hh9yZWeTmVM")
      true

  """
  def match?(url), do: String.match?(url, @regex)

  @doc """
  Parse and extract information from Youtube video URL.

  ## Examples

      iex> Videx.Youtube.parse("https://www.youtube.com/watch?v=Hh9yZWeTmVM")
      %Videx.Youtube{id: "Hh9yZWeTmVM", params: %{}, type: :long}

  """
  def parse(url) do
    case URI.parse(url) do
      %URI{host: "youtu.be", path: "/" <> id, query: query} ->
        %__MODULE__{type: :short, id: id, params: parse_params(query)}

      %URI{path: "/embed/" <> id, query: query} ->
        %__MODULE__{type: :embed, id: id, params: parse_params(query)}

      %URI{path: "/v/" <> id, query: query} ->
        %__MODULE__{type: :video, id: id, params: parse_params(query)}

      %URI{path: "/oembed", query: query} ->
        query
        |> parse_params()
        |> Map.pop("url")
        |> case do
          {nil, _} ->
            nil

          {original_url, params} ->
            case parse(original_url) do
              nil -> nil
              %{id: id} -> %__MODULE__{type: :oembed, id: id, params: params}
            end
        end

      %URI{query: query} ->
        query
        |> parse_params()
        |> Map.pop("v")
        |> case do
          {nil, _} ->
            nil

          {id, params} ->
            %__MODULE__{type: :long, id: id, params: params}
        end
    end
  end

  @doc """
  Generate URL for Youtube video.

  ## Examples

      iex> Videx.Youtube.generate(%Videx.Youtube{id: "Hh9yZWeTmVM"})
      "https://www.youtube.com/watch?v=Hh9yZWeTmVM"

      iex> Videx.Youtube.generate("Hh9yZWeTmVM", :short, %{t: 80})
      "https://youtu.be/Hh9yZWeTmVM?t=80"

  """
  def generate(_, type \\ :long, params \\ %{})

  def generate(%__MODULE__{id: id}, :long, params) do
    query =
      %{v: id}
      |> Map.merge(params)
      |> encode_query()

    %URI{
      scheme: "https",
      host: "www.youtube.com",
      path: "/watch",
      query: query
    }
    |> URI.to_string()
  end

  def generate(%__MODULE__{id: id}, :short, params) do
    %URI{
      scheme: "https",
      host: "youtu.be",
      path: "/#{id}",
      query: encode_query(params)
    }
    |> URI.to_string()
  end

  def generate(%__MODULE__{id: id}, :embed, params) do
    %URI{
      scheme: "https",
      host: "www.youtube.com",
      path: "/embed/#{id}",
      query: encode_query(params)
    }
    |> URI.to_string()
  end

  def generate(%__MODULE__{} = video, :oembed, params) do
    query =
      %{url: generate(video, :long)}
      |> Map.merge(params)
      |> encode_query()

    %URI{
      scheme: "https",
      host: "www.youtube.com",
      path: "/oembed",
      query: query
    }
    |> URI.to_string()
  end

  def generate(%__MODULE__{id: id}, :video, params) do
    %URI{
      scheme: "https",
      host: "www.youtube.com",
      path: "/v/#{id}",
      query: encode_query(params)
    }
    |> URI.to_string()
  end

  def generate(id, type, params) when is_binary(id) do
    generate(%__MODULE__{id: id}, type, params)
  end

  @doc """
  Generate embed HTML for Youtube video.

  ## Examples

      iex> Videx.Youtube.html(%Videx.Youtube{id: "Hh9yZWeTmVM"})
      "<iframe width=\\"560\\" height=\\"315\\" src=\\"https://www.youtube.com/embed/Hh9yZWeTmVM\\" frameborder=\\"0\\" allow=\\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\\" allowfullscreen></iframe>"

      iex> Videx.Youtube.html("Hh9yZWeTmVM", width: 320, height: 240)
      "<iframe width=\\"320\\" height=\\"240\\" src=\\"https://www.youtube.com/embed/Hh9yZWeTmVM\\" frameborder=\\"0\\" allow=\\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\\" allowfullscreen></iframe>"

  """
  def html(_, params \\ [])

  def html(%__MODULE__{id: id}, params) do
    width = params[:width] || 560
    height = params[:height] || 315
    start_at = params[:start_at]
    controls = is_nil(params[:controls]) || params[:controls]
    nocookie = params[:nocookie]

    host =
      case nocookie do
        true -> "www.youtube-nocookie.com"
        _ -> "www.youtube.com"
      end

    query =
      %{}
      |> Map.merge(
        if start_at do
          %{start: start_at}
        else
          %{}
        end
      )
      |> merge_query(:controls, controls, false)
      |> encode_query()

    url =
      %URI{
        scheme: "https",
        host: host,
        path: "/embed/#{id}",
        query: query
      }
      |> URI.to_string()

    "<iframe width=\"#{width}\" height=\"#{height}\" src=\"#{url}\" frameborder=\"0\" allow=\"accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture\" allowfullscreen></iframe>"
  end

  def html(id, params) when is_binary(id) do
    html(%__MODULE__{id: id}, params)
  end
end
